scripts/build_osm_network.py

```diff
@@ -595,131 +595,97 @@ def create_station_at_equal_bus_locations(
-            (
-                (
-                    bus_in_tol_epsg.geometry.distance(
-                        lines_epsgmod.geometry.loc[l].boundary.geoms[0]
-                    )
-                    > tol
-                )
-                | (
-                    bus_in_tol_epsg.geometry.distance(
-                        lines_epsgmod.geometry.loc[l].boundary.geoms[1]
-                    )
-                    > tol
-                )
-            )
+        # create perpendicular lines from the points that intersect with the line to the nearest points on the line
+        perpendicular_lines = [
+            LineString([point, nearest_point])
+            for point, nearest_point in zip(multi_points, nearest_points_list)
         ]
 
-        if not bus_in_tol_epsg.empty:
-            # add index of line to split
-            lines_to_split.append(l)
+        # split the line geom with the perpendicular lines using difference
+        split_line = line_geom.difference(MultiLineString(perpendicular_lines))
 
-            buses_locs = buses.geometry.loc[bus_in_tol_epsg.index]
+        # TODO: replace the end points of each line in the multistring with the points that intersect with the line
+        # unless there is a point that is intersecting the start point of the line
+        # in that case replace the start point with the point that is intersecting the start point of the line
 
-            # get new line geometries
-            new_geometries = _split_linestring_by_point(lines.geometry[l], buses_locs)
-            n_geoms = len(new_geometries)
+        # replace the line with the split line
+        df_l.loc[df_l["id"] == line_id, "geometry"] = split_line
 
-            # create temporary copies of the line
-            df_append = gpd.GeoDataFrame([lines.loc[l]] * n_geoms)
-            # update geometries
-            df_append["geometry"] = new_geometries
-            # update name of the line
-            df_append["line_id"] = [
-                str(df_append["line_id"].iloc[0]) + f"_{id}" for id in range(n_geoms)
-            ]
+    # explode the multilinestrings (not recommended, but included for completion)
+    # exploding the df should be done at the last step
+    # if an operation requires separate lines, it should be done using df.explode().apply(your_function)
+    # which is a lot more memory efficient
+    df_l = df_l.explode(index_parts=False)
 
-            lines_to_add.append(df_append)
-
-    if not lines_to_add:
-        return lines, buses
-
-    df_to_add = gpd.GeoDataFrame(pd.concat(lines_to_add, ignore_index=True))
-    df_to_add.set_crs(lines.crs, inplace=True)
-    df_to_add.set_index(lines.index[-1] + df_to_add.index, inplace=True)
+    # update line endings (included for completion, the scope of the function should be limited to fixing overpassing lines)
+    # commented out due to errors in the bus conversion function
+    # df_l = line_endings_to_bus_conversion(df_l)
```

Comment:
is this required? can we do the conversion after

Reply:
Yeah this function can be done as a last step to keep the values of the columns coherent